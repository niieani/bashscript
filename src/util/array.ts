/**
 * Combines two arrays by alternating their indices
 * combineAlternate([a, b, c], [1, 2, 3]) === [a, 1, b, 2, c, 3]
 */
export const combineAlternate = <T, Y>(
  arr1: Array<T>,
  arr2: Array<Y>,
  noLengthCheck = false,
  flipOrder = false,
): Array<T | Y> =>
  noLengthCheck || arr1.length >= arr2.length
    ? arr1.reduce(
        (combined, arr1value, index) =>
          index < arr2.length
            ? [
                ...combined,
                ...(flipOrder
                  ? [arr2[index], arr1value]
                  : [arr1value, arr2[index]]),
              ]
            : [...combined, arr1value],
        [] as Array<T | Y>,
      )
    : combineAlternate(arr2, arr1, true, true)

export const findLastIndex = <T>(
  predicate: (value: T, index: number, obj: T[]) => boolean,
) => (arr: ReadonlyArray<T>) => {
  const result = arr
    .slice()
    .reverse()
    .findIndex(predicate)
  return result >= 0 ? arr.length - 1 - result : result
}

export const last = <T>(array: Array<T>): T | undefined =>
  array.slice().reverse()[0]

export const ensureArray = <T>(maybeArray: T | Array<T>): Array<T> =>
  Array.isArray(maybeArray) ? maybeArray : [maybeArray]
