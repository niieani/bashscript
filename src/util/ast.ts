import * as AST from 'ts-simple-ast'
import * as acorn from 'acorn'
import {comment} from '../writer/syntax/comment'

export const getCommentObjects = (node: AST.Node) =>
  getComments(node).map((text) => comment(text, false))

export const getComments = ({compilerNode}: AST.Node) => {
  const commentPart = compilerNode
    .getFullText()
    .slice(0, compilerNode.getLeadingTriviaWidth())
  const comments: Array<acorn.Comment> = []
  acorn.parse(commentPart, {
    ranges: false,
    onComment: comments,
    allowHashBang: true,
  })
  return comments.map((comment) => `${comment.value.split('\n').join('\n#')}`)
}
