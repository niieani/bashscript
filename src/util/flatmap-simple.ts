export const flatmapSimple = <T>(nestedArray: Array<Array<T>>): Array<T> =>
  ([] as Array<T>).concat(...nestedArray)
