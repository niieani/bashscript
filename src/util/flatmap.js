// from https://gist.github.com/Kovensky/921c58de0c58be01aebf06d86cf9692f

if (typeof Array.prototype.flatMap !== 'function') {
  // eslint-disable-next-line no-extend-native
  Object.defineProperty(Array.prototype, 'flatMap', {
    enumerable: false,
    // https://bterlson.github.io/proposal-flatMap/#sec-Array.prototype.flatMap
    value: function flatMap (callback, thisArg = undefined) {
      const O = toObject(this)
      const A = arraySpeciesCreate(O, 0)
      // typo in the github pages render, it's missing
      // the ", 1" argument, but it is present in
      // the master source code
      flattenIntoArray(A, O, 0, 1, callback, thisArg)
      return A
    }
  })
}

if (typeof Array.prototype.flatten !== 'function') {
  // eslint-disable-next-line no-extend-native
  Object.defineProperty(Array.prototype, 'flatten', {
    enumerable: false,
    // https://bterlson.github.io/proposal-flatMap/#sec-Array.prototype.flatten
    value: function flatten (depth = Infinity) {
      const O = toObject(this)
      const A = arraySpeciesCreate(O, 0)
      depth = Number(depth)
      flattenIntoArray(A, O, 0, depth)
      return A
    }
  })
}

// https://bterlson.github.io/proposal-flatMap/#sec-FlattenIntoArray
function flattenIntoArray (target, source, start, depth, mapper, thisArg) {
  let targetIndex = start
  let sourceIndex = 0
  const sourceLen = toLength(source.length)

  while (sourceIndex < sourceLen) {
    const P = `${sourceIndex}`
    if (P in source) {
      let element = source[P]
      if (mapper) {
        element = mapper.call(thisArg, element, sourceIndex, target)
      }
      let spreadable
      // https://tc39.github.io/ecma262/#sec-isconcatspreadable
      if (typeof element !== 'object') {
        spreadable = false
      } else {
        spreadable = element[Symbol.isConcatSpreadable]
        if (typeof spreadable !== 'undefined') {
          spreadable = !!spreadable
        } else {
          spreadable = Array.isArray(element)
        }
      }

      // spec deviation: the spec specifies "depth >= 0",
      // but that incorrectly flattens one level too far.
      // https://github.com/bterlson/proposal-flatMap/issues/11
      if (spreadable && depth > 0) {
        const nextIndex = flattenIntoArray(target, element, targetIndex, depth - 1)
        // spec deviation: decrement targetIndex by 1 here
        // it will get incremented back at the tail of the loop
        // https://github.com/bterlson/proposal-flatMap/issues/13
        targetIndex = nextIndex - 1
      } else {
        if (targetIndex !== toLength(targetIndex)) {
          throw TypeError()
        }
        target[targetIndex] = element
      }
    }
    targetIndex += 1
    sourceIndex += 1
  }
  return targetIndex
}

// https://tc39.github.io/ecma262/#sec-arrayspeciescreate
function arraySpeciesCreate (original, length) {
  length = toLength(length)
  if (length < 0) {
    throw new RangeError()
  }

  let C
  if (!Array.isArray(original)) {
    C = original.constructor

    // Trick to implement the Realm check
    // from core-js
    if (typeof C === 'function' &&
      (C === Array || Array.isArray(C.prototype))) {
      C = undefined
    }

    if (typeof C === 'object' ? C !== null : typeof C === 'function') {
      C = C[Symbol.species]
      if (C === null) {
        C = undefined
      }
    }
  }

  if (typeof C === 'undefined') {
    C = Array
  }

  return new C(length)
}

// These are more-or-less like the core-js implementations
// in modules/_to-{length,integer,object}

function toLength (arg) {
  return arg > 0 ? Math.min(toInteger(arg), 0x1fffffffffffff) : 0
}

function toInteger (arg) {
  arg = +arg
  return isNaN(arg) ? 0 : (arg > 0 ? Math.floor : Math.ceil)(arg)
}

function toObject (arg) {
  if (arg === undefined) {
    throw new TypeError(`Can't call method on ${arg}`)
  }
  return Object(arg)
}
