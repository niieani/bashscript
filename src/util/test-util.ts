import {Project} from 'ts-morph'

export const makeTestFile = (fileContents: string) => {
  const project = new Project({
    useInMemoryFileSystem: true,
    addFilesFromTsConfig: false,
    compilerOptions: {
      strict: true,
      noLib: true,
    },
  })
  return project.createSourceFile('main.ts', fileContents)
}
