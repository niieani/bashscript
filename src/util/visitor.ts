import {ScopeContext} from '../writer/types'

export const passthrough = <T>(t: T) => t

export const getNameGenerator = (name: string) => <T>(
  wrapper: (t: string) => T = passthrough as (t: string) => T,
) => ({_context: {scopePath}}: ScopeContext) =>
  wrapper(`${scopePath.join('.')}.${name}`)
