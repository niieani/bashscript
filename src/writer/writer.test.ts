import '../util/flatmap'

type ScopeContext = {
  _context: TraverseState
} & TraverseState['scope']
type ScopeContextToAST<T = any> = (context : ScopeContext) => ASTResolvedExpression<T>

interface ASTObject<T = any> {
  type: ASTType
  data: T
  reduce: (context: TraverseState) => TraverseState
  parts?: Array<ASTExpression>
}

type ASTResolvedExpression<T = any> = string | ASTObject<T> | Array<string | ASTObject>
type ASTExpression = ScopeContextToAST | ASTResolvedExpression
type ASTList = Array<ASTExpression>

type ASTType =
  | 'root'
  | 'comment'
  | 'comment-inline'
  | 'declaration'
  | 'raw'
  | 'terminator'
  | 'starter'
  | 'lazy'
  | 'group'

// ? do we mark start of content flow (terminated by \n) vs inline content (a; b;)?
const starter: ASTObject<NoData> = {
  type: 'starter',
  data: {},
  reduce(context): TraverseState {
    const {processed, parts, indent} = context
    return {
      ...context,
      parts: indent ? [...parts, ''.padStart(indent)] : parts,
      processed: [...processed, this],
    }
  }
}

const findLastIndex = <T>(
  predicate: (value: T, index: number, obj: T[]) => boolean,
) => (arr: ReadonlyArray<T>) => {
  const result = arr.slice().reverse().findIndex(predicate)
  return result >= 0
    ? arr.length - 1 - result
    : result
}
const newLine = '\n'
const semi = '; '
const lastStarterAt = findLastIndex(({type}: ASTObject) => type === 'starter')
const lastTerminatorAt = findLastIndex(({type}: ASTObject) => type === 'terminator')
const contentSincePreviousTerminator = ({processed} : TraverseState) => {
  const previousTerminatorIndex = lastTerminatorAt(processed)
  return previousTerminatorIndex >= 0
    ? processed.slice(previousTerminatorIndex)
    : processed
}

type CommentData = {comment: string}
type NoData = {}

const getInlineComments = (objects: Array<ASTObject>) =>
  objects.filter(obj => obj.type === 'comment-inline') as Array<ASTObject<CommentData>>

function terminatorReduce(this: ASTObject<NoData>, context: TraverseState): TraverseState {
  const {processed, parts} = context
  const sinceLastTerminator = contentSincePreviousTerminator(context)
  const comments = getInlineComments(sinceLastTerminator)
  const commentText = `# ${comments.map(({data: {comment: text}}) => text).join(' | ')}`

  return {
    ...context,
    processed: [...processed, this],
    parts: comments.length
      ? [...parts, newLine, commentText] : [...parts, newLine],
  }
}

const terminator : ASTObject<NoData> = ({
  type: 'terminator',
  data: {},
  reduce: terminatorReduce,
})

const comment = (comment : string): ASTObject<{comment: string}> => ({
  reduce: defaultReduce,
  type: 'comment',
  parts: [`#${comment}`, terminator],
  data: {comment},
})

/**
 * Combines two arrays by alternating their indicies
 * combineAlternate([a, b, c], [1, 2, 3]) === [a, 1, b, 2, c, 3]
 */
const combineAlternate = <T, Y>(
  arr1 : Array<T>,
  arr2 : Array<Y>,
  noLengthCheck = false,
  flipOrder = false,
): Array<T | Y> =>
  (noLengthCheck || arr1.length >= arr2.length)
    ? arr1.reduce(
        (combined, arr1value, index) => (index < arr2.length)
          ? [...combined, ...(flipOrder ? [arr2[index], arr1value] : [arr1value, arr2[index]])]
          : [...combined, arr1value],
        [] as Array<T | Y>,
      )
    : combineAlternate(arr2, arr1, true, true)

// const prepareStatement = (nodes: Array<string | ASTObject>) => nodes
//   .filter(node => !isPureText(node) || node.length > 0)
//   .map(coerceStringToAST)
//   .concat(terminator)

// const resolvedStatement = (
//   strings: TemplateStringsArray,
//   ...parts: Array<ASTResolvedExpression>
// ) =>
//   prepareStatement(
//     combineAlternate(
//       Array.from(strings),
//       parts.flatten(1) as Array<string | ASTObject>
//     )
//   )

// const a = astGroup()

// const flattenObjectsReducer = (
//   strings: TemplateStringsArray,
//   ...parts: Array<ASTResolvedExpression>
// ) => (context: TraverseState) => ensureASTObject(
//   combineAlternate(Array.from(strings), parts.map((part) => ensureASTObject(part, context)))
//     .filter(node => !isPureText(node) || node.length > 0)
//     .map(coerceStringToAST)
//     .concat(terminator),
//   context,
// )

// const flattenedObjects = defineReducer({
//   reducer: flattenObjectsReducer(),
//   data:
// })

    // const x = [].flatMap(())
// const flattenPossible
// .map((part) => ensureASTObject(part))

// const statement = (
//   strings: TemplateStringsArray,
//   ...parts: Array<ASTObject<any> | string>,
// ) => combineAlternate(Array.from(strings), parts)
//   .filter(node => !isPureText(node) || node.length > 0)
//   .map(coerceStringToAST)
//   .concat(terminator)

const astGroup = (...children: ASTList): ASTObject<{children: ASTList}> => ({
  type: 'group',
  data: {children},
  reduce: function reduce(context: TraverseState): TraverseState {
    return reduceAST(children, context)
  },
})

/**
 * Template string for creating AST from nodes/strings and functions
 */
const ast = (
  strings: TemplateStringsArray,
  ...parts: Array<ASTExpression>,
): Array<ASTObject> => (
  parts.every(part => typeof part !== 'function')
    ? combineAlternate(
        Array.from(strings),
        parts.flatten(1) as Array<string | ASTObject>
      )
      .filter(node => !isPureText(node) || node.length > 0)
      .map(coerceStringToAST)
    : [
      astGroup(
        ({_context: context}) => ensureASTObject(
          combineAlternate(
            Array.from(strings).map(coerceStringToAST),
            parts.map((part) => ensureASTObject(part, context))
          ),
          context,
        )
      )
    ]
  )

const statement = (
  strings: TemplateStringsArray,
  ...parts: Array<ASTExpression>,
): Array<ASTObject> => ast(strings, ...parts).concat(terminator)

// const x = statement`declare ${variable}=${initializer}`

// comment helper:
// _COMMENT_=
// echo ${_COMMENT_# here is your comment}

function defaultReduce(this: ASTObject, context: TraverseState): TraverseState {
  const {processed, ...nextContext} = reduceAST(this.parts || [], context)
  return {
    ...nextContext,
    processed: [...processed, this],
  }
}

// in scopes { ... } and subshells ( ... ) we "fork" context
// i.e. ignore its result down the line
type DeclarationData = {variable: ASTObject, initializer: ASTObject | undefined}

type Reducer = (context: TraverseState) => TraverseState
type DefinedReducer<T> = Reducer & {data: T}

const defineReducer = <T>({reducer, data}: {
  reducer: (context: TraverseState) => TraverseState,
  data: T,
}) => Object.assign(reducer, {data})

const addToScope = (name: string) => defineReducer({
  data: {name},
  reducer: ({scope, parts, ...context}: TraverseState): TraverseState => {
    let safeName = name
    let append = 0
    while (safeName in scope) {
      append++
      // we need to rename this variable
      safeName = `${name}_${append}`
    }
    return {
      ...context,
      parts: [...parts, safeName],
      scope: {
        ...scope,
        [name]: {...scope.parent, toString: () => safeName, length: safeName.length},
      },
    }
  },
})

const declare = (
  variable : ASTObject<any>,
  initializer? : ASTObject<any>,
): ASTObject<DeclarationData> => ({
  parts: initializer
    ? statement`declare ${variable}=${initializer}`
    : statement`declare ${variable}`,
  type: 'declaration',
  data: {variable, initializer},
  reduce: defaultReduce,
})

const declareVariable = (
  name : string,
  initializer? : ASTObject<any>,
): ASTObject<DeclarationData> => declare(
  lazy(addToScope(name)),
  initializer,
)

const declareFunction = ({name, body}: {name: string, body: ASTExpression}) => ({
  type: 'function',
  parts: statement`function ${name} {
${body}
}`
})

const ex = statement`declare ${({name}) => name}`

const lazy = <T>(reduce: DefinedReducer<T>): ASTObject<T> => ({
  type: 'lazy',
  data: reduce.data,
  reduce,
})

const raw = (text : string): ASTObject<{text: string}> => ({
  reduce: function reduce({processed, parts, ...context}) {
    return {
      ...context,
      processed: [...processed, this],
      parts: [...parts, text],
    }
  },
  parts: [text],
  data: {text},
  type: 'raw',
})

const scopeHelper = (context : TraverseState) => ({
  ...context.scope,
  _context: context,
} as ScopeContext)

const isPureText = (textOrAST: ASTObject | string): textOrAST is string =>
  typeof textOrAST === 'string' || textOrAST.hasOwnProperty('toString')

const coerceStringToAST = <T>(textOrAST: ASTObject<T> | string) =>
  isPureText(textOrAST) ? raw(textOrAST) : textOrAST

const ensureASTObject = (node : ASTExpression, context : TraverseState): ASTObject =>
  Array.isArray(node)
    ? astGroup(...node.map(coerceStringToAST))
    : typeof node === 'function'
      ? ensureASTObject(node(scopeHelper(context)), context)
      : coerceStringToAST(node)

// const ast : ASTList = [
//   comment(`hello`),
// ]

interface TraverseState {
  parts: Array<string>
  processed: Array<ASTObject>
  scopePath: Array<string>
  parent: ASTObject
  indent: number,
  scope: {
    [variableName: string]: ASTObject & {toString(): string, length: number},
  }
}

const astRoot: ASTObject<NoData> = {
  type: 'root',
  data: {},
  reduce: (ctx) => ctx,
}

const emptyContext: TraverseState = {
  parts: [], processed: [], scopePath: [], scope: {}, parent: astRoot,
  indent: 0,
}

const reduceAST = (ast : ASTList, context : TraverseState = emptyContext): TraverseState =>
  ast.reduce(
    (context : TraverseState, node) => ensureASTObject(
      node,
      context,
    ).reduce(context),
    context
  )

const print = (ast : ASTList, context : TraverseState = emptyContext) =>
  reduceAST(ast, context).parts.join('')

describe('writer', () => {
  test('comment', () => {
    expect(
      print([
        comment('testing')
      ])
    ).toMatchSnapshot()
  })

  test('terminator', () => {
    expect(
      print([
        terminator
      ])
    ).toMatchSnapshot()
  })

  test('declaration', () => {
    expect(
      print([
        declareVariable('hello', raw('test'))
      ])
    ).toMatchSnapshot()
  })

  test('statement', () => {
    expect(
      print(
        statement`echo test`
      )
    ).toMatchSnapshot()
  })

  test('statement with raw', () => {
    expect(
      print(
        statement`echo ${raw('raw test')}`
      )
    ).toMatchSnapshot()
  })

  test('scope is populated by variables', () => {
    expect(
      print([
        declareVariable('example'),
        ({example}) => statement`echo ${example}`,
      ])
    ).toMatchSnapshot()
  })

  test('conflicting variable names are renamed', () => {
    expect(
      print([
        declareVariable('example'),
        ({example}) => statement`echo ${example}`,
        declareVariable('example'),
        ({example}) => statement`echo ${example}`, // will be example_1
      ])
    ).toMatchSnapshot()
  })
})

describe('ast', () => {
  test('comment', () => {
    expect(
      reduceAST([comment('testing')])
    ).toMatchSnapshot()
  })

  describe('ensureASTObject', () => {
    test('[] => group', () => {
      expect(
        ensureASTObject([], emptyContext)
      ).toHaveProperty('type', 'group')
    })
    test('[text] => group with one raw node', () => {
      const result = ensureASTObject(['text'], emptyContext)
      expect(result).toHaveProperty('type', 'group')
      expect(result.data).toHaveProperty('children')
      expect(result.data.children).toHaveLength(1)
      expect(result.data.children[0]).toHaveProperty('type', 'raw')
    })
    test('text => one raw node', () => {
      const result = ensureASTObject('text', emptyContext)
      expect(result).toHaveProperty('type', 'raw')
      expect(result.data).toHaveProperty('text')
      expect(result.data.text).toBe('text')
    })
    test('() => [text] => group with one raw node', () => {
      const result = ensureASTObject(() => ['text'], emptyContext)
      expect(result).toHaveProperty('type', 'group')
      expect(result.data).toHaveProperty('children')
      expect(result.data.children).toHaveLength(1)
      expect(result.data.children[0]).toHaveProperty('type', 'raw')
    })
    test('() => text => one raw node', () => {
      const result = ensureASTObject(() => 'text', emptyContext)
      expect(result).toHaveProperty('type', 'raw')
      expect(result.data).toHaveProperty('text')
      expect(result.data.text).toBe('text')
    })
    test('() => statement => 2 nodes', () => {
      const result = ensureASTObject(() => statement`text`, emptyContext)
      expect(result).toHaveProperty('type', 'group')
      expect(result.data).toHaveProperty('children')
      expect(result.data.children).toHaveLength(2)
    })
  })
})

describe('utils', () => {
  test('combineAlternate', () => {
    expect(
      combineAlternate(['a', 'b', 'c'], [1, 2, 3])
    ).toMatchObject(['a', 1, 'b', 2, 'c', 3])
  })
})
