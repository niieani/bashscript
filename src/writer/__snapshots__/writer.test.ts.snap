// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ast comment 1`] = `
Object {
  "parent": Object {
    "data": Object {},
    "reduce": [Function],
    "type": "root",
  },
  "parts": Array [
    "#testing",
    "
",
  ],
  "processed": Array [
    Object {
      "data": Object {
        "text": "#testing",
      },
      "parts": Array [
        "#testing",
      ],
      "reduce": [Function],
      "type": "raw",
    },
    Object {
      "data": Object {},
      "reduce": [Function],
      "type": "terminator",
    },
    Object {
      "data": Object {
        "comment": "testing",
      },
      "parts": Array [
        "#testing",
        Object {
          "data": Object {},
          "reduce": [Function],
          "type": "terminator",
        },
      ],
      "reduce": [Function],
      "type": "comment",
    },
  ],
  "scope": Object {},
  "scopePath": Array [],
}
`;

exports[`writer comment 1`] = `
"#testing
"
`;

exports[`writer conflicting variable names are renamed 1`] = `
"declare example
echo example
declare example_1
echo example_1
"
`;

exports[`writer declaration 1`] = `
"declare hello=test
"
`;

exports[`writer scope is populated by variables 1`] = `
"declare example
echo example
"
`;

exports[`writer statement 1`] = `
"echo test
"
`;

exports[`writer statement with raw 1`] = `
"echo raw test
"
`;

exports[`writer terminator 1`] = `
"
"
`;
